### ISSUE TITLE
Duplicate resources when using implementation-specific annotations

### ISSUE THREAD

**Issue by pottekkat** [kubernetes-sigs/ingress2gateway#109]
**What happened**:

When converting Ingress resources with implementation-specific annotations, the output contains duplicated `Gateway` and `HTTPRoute` resources.

See https://github.com/kubernetes-sigs/ingress2gateway/pull/104#issuecomment-1856463361 for details.

Also see: https://github.com/kubernetes-sigs/ingress2gateway/pull/104#issuecomment-1857538831 and https://github.com/kubernetes-sigs/ingress2gateway/pull/104#issuecomment-1857580813

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

See https://github.com/kubernetes-sigs/ingress2gateway/pull/104#issuecomment-1856463361 for reproducing.

**Anything else we need to know?**:

No.

**Comment by pottekkat** [comment]
From a quick look, the issue seems to be because each provider calls the common `ToGateway` function. If that is the case, instead of changing the provider logic, we can just disable all providers by default and then user can enable any one provider, which would usually be the case.

**Comment by mlavacca** [comment]
> From a quick look, the issue seems to be because each provider calls the common `ToGateway` function. If that is the case, instead of changing the provider logic, we can just disable all providers by default and then user can enable any one provider, which would usually be the case.

Similarly to what @levikobi suggested in their comment, I suggest adding the `IngressClass` name to the common converter function signature and letting the common machinery filter out resources for not matching ingress classes. @levikobi does it make sense to you?

**Comment by levikobi** [comment]
> > From a quick look, the issue seems to be because each provider calls the common `ToGateway` function. If that is the case, instead of changing the provider logic, we can just disable all providers by default and then user can enable any one provider, which would usually be the case.
> 
> Similarly to what @levikobi suggested in their comment, I suggest adding the `IngressClass` name to the common converter function signature and letting the common machinery filter out resources for not matching ingress classes. @levikobi does it make sense to you?

Absolutely @mlavacca, that's exactly what I meant 👍

**Comment by pottekkat** [comment]
In the provider code, we get all the gateway resources from the common `ToGateway` and the parse it and add implementation specific configuration to those resources which have custom annotations and return everything back. If the provider only gets a filtered gateway resources, then we will only return resources which has converted annotations? The rest will be filtered out because they aren't implementation specific?

Maybe I am missing something.

**Comment by levikobi** [comment]
@pottekkat pretty much yes, but after reading your comment I'm realizing it isn't as intuitive as it should be.

Maybe we should add a new function to the common package, which will filter ingresses based on class name. Every provider will call it before calling ToGateway.
It will be more understandable and also keeps ToGateway function with one responsibility.

Thoughts @mlavacca @pottekkat?

**Comment by levikobi** [comment]
I tend to agree with you @mlavacca, that's a good point IMO. I would also prefer not to let the implementations the extra "power" / "responsibility". Maybe we can call the filter function in the i2gw pkg, and only pass implementations their associated resources?

**Comment by LiorLieberman** [comment]
Thanks for flagging this. I think it is touching a topic we discussed on https://github.com/kubernetes-sigs/ingress2gateway/pull/99 with @dpasiukevich. More specifically this comment - https://github.com/kubernetes-sigs/ingress2gateway/pull/99#issuecomment-1833859435.

The solution to this would be to also move ingresses to be fetched by the provider (perhaps every provider will query the ingresses with the relevant ingress classes it needs). Same as we do with CRDs. 

Thoughts?
Any cons I did no think of if we are proceeding with this change?
